### PR TITLE
Move GeometryCollection to separate file

### PR DESCRIPTION
**Pull request by alexfikl**
~~(also contains #140, so waiting for that to go in first)~~

* Moved `GeometryCollection` to `pytential.collection` and added `GeometryLike` for typing annotations.
* Moved `DOFDescriptor` to `pytential.symbolic.dof_desc` and added `DOFDescriptorLike` for type annotations.

### CODE REVIEW

File: pytential/collection.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+__doc__ = """
+.. class:: GeometryLike
+
+    Types accepted by the :class:`GeometryCollection`.
+
+.. class:: DOFDescriptorLike
+
+    Types convertible to a :class:`~pytential.symbolic.primitives.DOFDescriptor`
+    by :func:`~pytential.symbolic.primitives.as_dofdesc`.
```
Comment (alexfikl) [inline pytential/collection.py:44]: Move this to `primitives.py`?

Comment (inducer) [reply pytential/collection.py:44]: `primitives` is already a bit long (2k lines or so). Move everything `DOFDesc`-y to `dof_desc.py`?

Comment (alexfikl) [reply pytential/collection.py]: Moved it in b6c3c7a and added some types.

There shouldn't be any functionality changes in there, but the diff sure looks scary :cry:

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**